future-apps/agent/contracts/SignatureValidator.sol

```diff
@@ -1,36 +1,74 @@
 pragma solidity 0.4.24;
 
-// From github.com/DexyProject/protocol
-// + https://github.com/0xProject/0x-monorepo/blob/master/packages/contracts/contracts/protocol/Exchange/MixinSignatureValidator.sol
+// Inspired in https://github.com/horizon-games/multi-token-standard/blob/master/contracts/utils/SignatureValidator.sol
 // This should probably be moved into aOS: https://github.com/aragon/aragonOS/pull/442
 
+import "./standards/ERC1271.sol";
+
 
 library SignatureValidator {
     enum SignatureMode {
-        Invalid,
-        EIP712,
-        EthSign
+        Invalid, // 0x00
+        EIP712,  // 0x01
+        EthSign, // 0x02
+        ERC1271, // 0x03
+        NMode    // 0x04, to check if mode is specified, leave at the end
     }
 
+    // bytes4(keccak256("isValidSignature(bytes,bytes)")
+    bytes4 constant public ERC1271_RETURN_VALID_SIGNATURE = 0x20c13b0b;
+    uint256 internal constant ERC1271_ISVALIDSIG_MAX_GAS = 250000;
+
     /// @dev Validates that a hash was signed by a specified signer.
     /// @param hash Hash which was signed.
     /// @param signer Address of the signer.
-    /// @param signature ECDSA signature along with the mode (0 = Invalid, 1 = EIP712, 2 = EthSign) {mode}{r}{s}{v}.
+    /// @param signature ECDSA signature along with the mode (0 = Invalid, 1 = EIP712, 2 = EthSign, 3 = ERC1271) {mode}{r}{s}{v}.
     /// @return Returns whether signature is from a specified user.
-    function isValidSignature(bytes32 hash, address signer, bytes signature) internal pure returns (bool) {
+    function isValidSignature(bytes32 hash, address signer, bytes signature) internal view returns (bool) {
         if (signature.length == 0) {
```

Comment:
This should be `<= 1` since mode is always required

Reply:
`== 0` == `<= 1`, I think I don't understand what you mean here

Reply:
Length being 1 is valid for ERC1271 mode signatures. We could perform the check before sending it over to the erc1271 contract, but that contract may have logic which allows empty signatures